### DISCUSSION TITLE
Gentoo Updates

### DISCUSSION THREAD

**Discussion by immolo** [frno7/mipsr5900el-gentoo-linux-gnu#18] [Show and tell]
Been a while however I have been doing some work in the background with MIPS on Gentoo to make the process a lot smoother.

My main project has been getting the [Catalyst](https://wiki.gentoo.org/wiki/Project:Catalyst)t build system to build mipsel3 musl stage3 files which has been successful and awaiting a PR into the main project.

If you are interested in the meantime then I have a write at https://wiki.gentoo.org/wiki/Cataylst_Musl_Stages_Creation which takes you through all the steps, obviously the tuple needs to be changed here for the PS2. There is also https://wiki.gentoo.org/wiki/User:Immolo/GentooPS2 which needs a little more work before it usable but it will provide helpful hints with things I have discovered.

I may release an update to the LiveUSB using the new codebase however I'll most likely just wait until hard drive support is added to the kernel so Gentoo is actually worth using on the PS2.

**Comment by AKuHAK** [comment]
HDD support isnt that easy, so I advice to not wait for it. BTW where you post your LiveUsb images?

**Reply by immolo** [reply]
http://www.immolo.co.uk/gentoo-ps2-liveusb-alpha1.tar.xz

Not really much point rebasing it right now but I'll assess it when the need arises.

**Comment by frno7** [comment]
Ooh, that’s very nice! :-) Maybe I’m aiming a bit high with multi-buffer hard disk read/write DMA for a first implementation. Performance would be quite good though, as mentioned in https://github.com/frno7/linux/issues/18#issuecomment-880095988.

> ```sh
> ...
> modprobe hid-generic
> sleep 5
> 
> #Mount USB
> mount -t ext4 -o ro -L PS2Root /mnt || echo "Failed to mount root device"
> sleep 30
> ...
> ```

Are the `sleep` in `initramfs/init` for slowing down kernel prints, to make them easier to read? Things that genuinely need delays?

**Reply by immolo** [reply]
After dealing with the USB1.1 speeds, I dream of 31MB/s ;)

Not having a sleep causes some issues where it doesn't always load the USB drive in time, 30 is a little high but it seems to catch everyone's PS2 to allow booting. That said using `read` is supposed to be the correct way to do this, it's just further down the list of things to look into as making sure the MIPS 2GB memory limit isn't hit randomly when building GCC in qemu-user is where more focus is needed from me first.

**Reply by frno7** [reply]
Would that be `read` with some kind of `while` loop in the shell, testing whether the drive has become available?

> ... making sure the MIPS 2GB memory limit isn't hit randomly when building GCC in qemu-user is where more focus is needed from me first.

The o32 ABI is 32-bit throughout. The R5900 can do n32 too, in principle, but sadly pointers are still 32-bit, so the 2 GiB limit still apply, although some integer types etc. become 64-bit (similar to the [x32 ABI](https://en.wikipedia.org/wiki/X32_ABI) for the x86). I suppose GCC could be cross-compiled instead, so QEMU could be avoided for compiling GCC itself?

I’ve updated the kernel to v5.4.221 just now, by the way.

**Reply by immolo** [reply]
> Would that be `read` with some kind of `while` loop in the shell, testing whether the drive has become available?

I was told it should just wait for the device but I've still not got around to looking into it.

> > ... making sure the MIPS 2GB memory limit isn't hit randomly when building GCC in qemu-user is where more focus is needed from me first.
> 
> The o32 ABI is 32-bit throughout. The R5900 can do n32 too, in principle, but sadly pointers are still 32-bit, so the 2 GiB limit still apply, although some integer types etc. become 64-bit (similar to the [x32 ABI](https://en.wikipedia.org/wiki/X32_ABI) for the x86). I suppose GCC could be cross-compiled instead, so QEMU could be avoided for compiling GCC itself?

I solved it with a package.env for gcc to remove `-pipe` and that keeps it under the limit for now but I've got some LDFLAG tricks with `-no-keep-memory` if it comes back. I wish I could use crossdev with Catalyst but that day is still a way off however there is work being done on this.

> I’ve updated the kernel to v5.4.221 just now, by the way.

That's great news and I can't wait to play with it once I've finished with the projects I've got on on my TODO list :)

**Reply by immolo** [reply]
Also I forgot to add, n32 isn't officially support by Musl so while I know it builds, I need to do further testing to see how viable it is to support.

**Reply by frno7** [reply]
Right. GNU libc would need a patch as well, as I recall, for R5900 specifically. An n32 kernel would be able to run o32 programs too, with the benefit of having a greater chance of managing the 128-bit multimedia instructions of the R5900 (issue [#5](https://github.com/frno7/linux/issues/5)). However, considering that the kernel most likely will [sunset the x32 ABI](https://lkml.org/lkml/2018/12/10/1151), general kernel MIPS support for n32 might eventually deteriorate badly, in which case upgrading (successfully) to recent kernels would become progressively more difficult. This has implications on issue [#1](https://github.com/frno7/linux/issues/1), but one can of course stay on a 5.4 or some other decent kernel version more or less indefinitely. :-)

**Reply by immolo** [reply]
I was explained it as n32 is basically the MIPS version on x32 on x86_64, I was also told it isn't that useful either but I'm still new to MIPS so I would recommend #gentoo-mips for someone with more knowledge to explain.

On a personal note though it will give me a way to test n32 bugs so at least one person will find it useful :)

**Reply by frno7** [reply]
The R5900 hardware won’t do better than n32, meaning any kind of proper o64 kernel support for the R5900 has been ruled out by MIPS kernel maintainers. So n32 is a sweet spot for the R5900, the best it can offer, but admittedly weird otherwise. A [suitably patched up QEMU](https://patchwork.kernel.org/project/qemu-devel/cover/cover.1541701393.git.noring@nocrew.org/) can manage n32 for the R5900.

**Reply by immolo** [reply]
I meant it that it suppose to give 64bit instructions to a 32bit system so might not be worth it.

Don't rush it though as there are some mips profile bugs in Gentoo currently stopping me playing with my new toy (creator ci40) and I think this might show up here but I'd need to confirm that.

**Reply by frno7** [reply]
The 64-bit and, especially, the 128-bit (multimedia) registers are defining features of the R5900, so there are good reasons to make them usable. Thus n32... :-)

**Reply by immolo** [reply]
I should know better than to doubt your knowledge :P

In that case I'm looking forward to all the Musl n32 bugs to try and fix.

**Reply by frno7** [reply]
Well, you might want to wait for the kernel to catch up on the n32. We’re not there yet. :-)

**Reply by immolo** [reply]
Don't worry I've got enough on my todo list to keep me busy. 

That reminds me I need to check on your musl patch soon,

**Reply by AKuHAK** [reply]
> However, considering that the kernel most likely will sunset the x32 ABI, general kernel MIPS support for n32 might eventually deteriorate badly, in which case upgrading (successfully) to recent kernels would become progressively more difficult.

Does that mean, that your work will never be merged upstream as considered obsolete for Linux team?

**Reply by frno7** [reply]
> Does that mean, that your work will never be merged upstream as considered obsolete for Linux team?

That seems to be the case, most likely, for several reasons:

- MIPS Linux kernel maintainer review capacity has been declining sharply for several years, and 10,000+ lines of code for PlayStation 2 support is apparently much too much code to consider for inclusion;
- MIPS hardware, software, and interest (in general) have been declining for several years, meaning especially software related to MIPS tend to break more often;
- vintage hardware (in general) and their related subsystems, are being purged from the Linux kernel, for the benefit of expanding and simplifying maintenance of modern hardware.

Staying on a stable kernel version (5.4, for instance) indefinitely has some advantages. So, why not? :-)

**Reply by immolo** [reply]
> > Does that mean, that your work will never be merged upstream as considered obsolete for Linux team?
> 
> That seems to be the case, most likely, for several reasons:
> 
>     * MIPS Linux kernel maintainer review capacity has been declining sharply for several years, and 10,000+ lines of code for PlayStation 2 support is apparently much too much code to consider for inclusion;
> 
>     * MIPS hardware, software, and interest (in general) have been declining for several years, meaning especially software related to MIPS tend to break more often;
> 
>     * vintage hardware (in general) and their related subsystems, are being purged from the Linux kernel, for the benefit of expanding and simplifying maintenance of modern hardware.
> 
> 
> Staying on a stable kernel version (5.4, for instance) indefinitely has some advantages. So, why not? :-)

That's huge shame they won't as it makes getting the PS2 an official Gentoo port a harder sell for me :P That said it's pretty easy to workaround old versions when using Musl as a libc.

**Reply by frno7** [reply]
> That's huge shame they won't as it makes getting the PS2 an official Gentoo port a harder sell for me :P

The `-mfix-r5900` option was a _very_ good step forward, as it means we’ve got official Gentoo/MIPS compatibility. Gentoo supplies downloads for hardware in even more dire situations than the PlayStation 2, I should think. For instance HP-PA and IA-64?

> That said it's pretty easy to workaround old versions when using Musl as a libc.

Hmm, work around old kernel versions? Old Gentoo versions? Please elaborate. :-)

**Reply by immolo** [reply]
It's not impossible but from experience they aren't as keen.

Old kernel versions on latest Gentoo, I've done a lot of work on this topic in the past and even have a guide for going back as far realistically usable (2.6.39 btw).

https://wiki.gentoo.org/wiki/User:Immolo/OldKernelInstall if you want to see how it is done.

**Reply by frno7** [reply]
> It's not impossible but from experience they aren't as keen.

That’s understandable. Would people want to upgrade regularly to recent kernels, if the PlayStation 2 port actually was officially merged? Security fixes exempted, upgrading doesn’t seem very compelling, since newer kernels often mean loss of features, lower performance, and upgrade breakages, compared to older ones? Has the kernel had any new features in, say, the last few years, that would make upgrading meaningful?

These questions may be difficult to fully answer, but to me, playing with retro/hobby hardware is much more worthwhile than struggling with kernel deprecation policies. :-)

> Old kernel versions on latest Gentoo, I've done a lot of work on this topic in the past and even have a guide for going back as far realistically usable (2.6.39 btw). https://wiki.gentoo.org/wiki/User:Immolo/OldKernelInstall if you want to see how it is done.

Heh, that kernel is 10+ years old. Sure, systemd, pam, etc. may need to be replaced.

**Reply by immolo** [reply]
> > It's not impossible but from experience they aren't as keen.
> 
> That’s understandable. Would people want to upgrade regularly to recent kernels, if the PlayStation 2 port actually was officially merged? Security fixes exempted, upgrading doesn’t seem very compelling, since newer kernels often mean loss of features, lower performance, and upgrade breakages, compared to older ones? Has the kernel had any new features in, say, the last few years, that would make upgrading meaningful?
> 
> These questions may be difficult to fully answer, but to me, playing with retro/hobby hardware is much more worthwhile than struggling with kernel deprecation policies. :-)

Judging by the Wii being stuck on 3.10.5 I would assume not, I think the last change was around 4.14 where some core applications started relying on a feature although GLIBC can be a pain rasing the minimum kernel version something. I would have to look deeper though as these are just off my head.

> > Old kernel versions on latest Gentoo, I've done a lot of work on this topic in the past and even have a guide for going back as far realistically usable (2.6.39 btw). https://wiki.gentoo.org/wiki/User:Immolo/OldKernelInstall if you want to see how it is done.
> 
> Heh, that kernel is 10+ years old. Sure, systemd, pam, etc. may need to be replaced.

Had to make it fun somehow :)
